modules/packages/arch/metadata.go

```diff
@@ -0,0 +1,330 @@
+	Depends           []string
+	OptDepends        []string
+	MakeDepends       []string
+	CheckDepends      []string
+	Backup            []string
+}
+
+// Function that recieves arch package archive data and returns it's metadata.
+func EjectMetadata(filename, domain string, pkg []byte) (*Metadata, error) {
+	pkgreader := io.LimitReader(bytes.NewReader(pkg), 250000)
+	var buf bytes.Buffer
+	err := archiver.DefaultZstd.Decompress(pkgreader, &buf)
+	if err != nil {
+		if !errors.Is(err, io.ErrUnexpectedEOF) {
+			return nil, err
+		}
+	}
+	splt := strings.Split(buf.String(), "PKGINFO")
+	if len(splt) < 2 {
+		return nil, errors.New("unable to eject .PKGINFO from archive")
+	}
+	raw := splt[1][0:10000]
+	inssize := int64(len(pkg))
+	compsize := ejectInt64(raw, "size")
+	unixbuilddate := ejectInt64(raw, "builddate")
+	return &Metadata{
+		Filename:          filename,
+		Name:              ejectString(raw, "pkgname"),
+		Base:              ejectString(raw, "pkgbase"),
+		Version:           ejectString(raw, "pkgver"),
+		Description:       ejectString(raw, "pkgdesc"),
+		CompressedSize:    inssize,
+		CompressedSizeMib: ByteCountSI(inssize),
+		InstalledSize:     compsize,
+		InstalledSizeMib:  ByteCountSI(compsize),
+		MD5:               md5sum(pkg),
+		SHA256:            sha256sum(pkg),
+		URL:               ejectString(raw, "url"),
+		BuildDate:         unixbuilddate,
+		BuildDateStr:      ReadableTime(unixbuilddate),
+		BaseDomain:        domain,
+		Packager:          ejectString(raw, "packager"),
+		Provides:          ejectStrings(raw, "provides"),
+		License:           ejectStrings(raw, "license"),
+		Arch:              ejectStrings(raw, "arch"),
+		Depends:           ejectStrings(raw, "depend"),
+		OptDepends:        ejectStrings(raw, "optdepend"),
+		MakeDepends:       ejectStrings(raw, "makedepend"),
+		CheckDepends:      ejectStrings(raw, "checkdepend"),
+		Backup:            ejectStrings(raw, "backup"),
+	}, nil
+}
+
+func ejectString(raw, field string) string {
+	splitted := strings.Split(raw, "\n"+field+" = ")
+	if len(splitted) < 2 {
+		return ``
+	}
+	return strings.Split(splitted[1], "\n")[0]
+}
+
+func ejectStrings(raw, field string) []string {
+	splitted := strings.Split(raw, "\n"+field+" = ")
+	if len(splitted) < 2 {
+		return nil
+	}
+	var rez []string
+	for i, v := range splitted {
+		if i == 0 {
+			continue
+		}
+		rez = append(rez, strings.Split(v, "\n")[0])
+	}
+	return rez
+}
+
+func ejectInt64(raw, field string) int64 {
+	splitted := strings.Split(raw, "\n"+field+" = ")
+	if len(splitted) < 2 {
+		return 0
+	}
+	i, err := strconv.ParseInt(strings.Split(splitted[1], "\n")[0], 10, 64)
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func ByteCountSI(b int64) string {
+	const unit = 1000
+	if b < unit {
+		return fmt.Sprintf("%d B", b)
+	}
+	div, exp := int64(unit), 0
+	for n := b / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f %cB", float64(b)/float64(div), "kMGTPE"[exp])
+}
```

Comment:
The fields which use this function are unused. If there must be a human readable display of the size, the conversion should be done in UI.